### PR TITLE
Enabling TLS

### PR DESCRIPTION
**Pull request by alesstimec**
Initial implementation of a relation to easyrsa to
retrieve server certs and put them in JKSs.

### CODE REVIEW

File: charm/kafka/reactive/kafka.py [hunk 1]
```diff
@@ -166,7 +185,8 @@ def storage_attach():
     log_dir = os.path.join(mount, "logs")
     unitdata.kv().set('kafka.storage.log_dir', log_dir)
     hookenv.log('Kafka logs storage attached at {}'.format(log_dir))
-    # Stop Kafka; removing the kafka.started state will trigger a reconfigure if/when it's ready
+    # Stop Kafka; removing the kafka.started state will trigger \
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py]: Could remove the trailing ` \` for a comment.

File: charm/kafka/templates/server.properties [hunk 2]
```diff
@@ -28,12 +28,20 @@ broker.id={{ broker_id }}
...
 #     listeners = PLAINTEXT://your.host.name:9092
-listeners=PLAINTEXT://{{ bind_addr }}:{{ port }}
+listeners=SSL://{{ bind_addr }}:{{ port }}
+
+ssl.truststore.location={{ ca_keystore }}
+ssl.truststore.password={{ keystore_password }}
+ssl.keystore.location={{ server_keystore }}
+ssl.keystore.password={{ keystore_password }}
+ssl.key.password={{ keystore_password }}
+security.inter.broker.protocol=SSL
```
Comment (cmars) [inline charm/kafka/templates/server.properties:38]: Do the brokers in a cluster need to issue client certs in order to be able to talk to each other?

**Review by cmars — COMMENTED** [review]
I think the kafka unit will also need to issue itself a client cert so that the kafka actions can connect to the TLS endpoint.

File: charm/kafka/reactive/kafka.py (hunk 1 above)
Comment (alesstimec) [reply charm/kafka/reactive/kafka.py]: done

File: charm/kafka/templates/server.properties (hunk 2 above)
Comment (alesstimec) [reply charm/kafka/templates/server.properties:38]: no.. easyrsa will issue client certs

File: charm/kafka/lib/charms/layer/kafka.py [hunk 3]
```diff
@@ -145,3 +173,11 @@ def get_ip_for_interface(self, network_interface):
                     u"This machine does not have an interface '{}'".format(
                         network_interface))
             return ip
+
```
Comment (cmars) [inline charm/kafka/lib/charms/layer/kafka.py:176]: Add extra newline (pep8).

File: charm/kafka/reactive/kafka.py [hunk 4]
```diff
@@ -184,3 +208,168 @@ def storage_detaching():
...
+            'client.crt',
+        ),
+        key_path=os.path.join(
+            KAFKA_SNAP_DATA,
+            'client.key'
+        )
+    )
+
+
+# TODO we need to restart server if we change keystore
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py]: Maybe something like:
```
@when_file_changed(os.path.join(KAFKA_SNAP_DATA, "kafka.server.jks"))
@when_file_changed(os.path.join(KAFKA_SNAP_DATA, "kafka.client.jks"))
def restart_when_keystore_changed():
  Kafka().restart()
```

**Review by cmars — APPROVED** [review]
LGTM! Couple of suggestions.

File: charm/kafka/lib/charms/layer/kafka.py (hunk 3 above)
Comment (alesstimec) [reply charm/kafka/lib/charms/layer/kafka.py:176]: added

File: charm/kafka/reactive/kafka.py (hunk 4 above)
Comment (alesstimec) [reply charm/kafka/reactive/kafka.py]: done

**Pull request merged**